# 01 Dataset Understanding

This notebook performs the initial dataset understanding and profiling for **Problem 6: The Overpayment Signal**.

### Objectives:
1. Load and inspect the shape, schema, and sample records of the `cases.csv` and `payments.csv` datasets.
2. Profile column statistics, missing values, duplicates, and ranges.
3. Investigate the relationships between the two datasets.
4. Identify potential demographic fields for future fairness analysis.
5. Document initial observations directly supported by the data.

---

## 1. Import Required Libraries

We import standard packages for data loading, analysis, and visualization.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set display options for better visibility
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)
sns.set_theme(style='whitegrid')

## 2. Load Dataset Files

We load the raw CSV files `cases.csv` and `payments.csv` from the `data/raw/` directory.

In [2]:
df_cases = pd.read_csv('../data/raw/cases.csv')
df_payments = pd.read_csv('../data/raw/payments.csv')
print('Datasets loaded successfully!')

Datasets loaded successfully!


## 3. Dataset Shapes

We print the number of rows and columns for both datasets.

In [3]:
print(f'Cases Dataset Shape: {df_cases.shape[0]} rows, {df_cases.shape[1]} columns')
print(f'Payments Dataset Shape: {df_payments.shape[0]} rows, {df_payments.shape[1]} columns')

Cases Dataset Shape: 4200 rows, 13 columns
Payments Dataset Shape: 24756 rows, 6 columns


## 4. Column Names and Data Types

We inspect the column names, counts of non-null values, and data types for both datasets.

In [4]:
print('--- CASES INFO ---')
df_cases.info()
print('\n--- PAYMENTS INFO ---')
df_payments.info()

--- CASES INFO ---
<class 'pandas.DataFrame'>
RangeIndex: 4200 entries, 0 to 4199
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   case_id              4200 non-null   str    
 1   district             4200 non-null   str    
 2   household_size       4200 non-null   int64  
 3   age_band             4200 non-null   str    
 4   language_preference  4200 non-null   str    
 5   tenure               4200 non-null   str    
 6   opened_date          4200 non-null   str    
 7   status               4200 non-null   str    
 8   closure_month        251 non-null    str    
 9   monthly_award        4200 non-null   float64
 10  payment_adjustments  4200 non-null   int64  
 11  contact_attempts     4200 non-null   int64  
 12  months_since_review  4200 non-null   int64  
dtypes: float64(1), int64(4), str(8)
memory usage: 426.7 KB

--- PAYMENTS INFO ---
<class 'pandas.DataFrame'>
RangeIndex: 24756 entries

## 5. Sample Records

We display the first 5 records of each dataset to check the data values directly.

In [5]:
print('--- CASES HEAD ---')
display(df_cases.head())
print('\n--- PAYMENTS HEAD ---')
display(df_payments.head())

--- CASES HEAD ---


,case_id,district,household_size,age_band,language_preference,tenure,opened_date,status,closure_month,monthly_award,payment_adjustments,contact_attempts,months_since_review
0,C-30000,Calder Central,2,45-59,English,Private tenancy,2024-11-13,Active,NaN,1026.76,0,3,0
1,C-30001,Weybridge,1,30-44,English,Private tenancy,2023-10-21,Active,NaN,497.45,0,0,0
2,C-30002,Northgate,2,45-59,English,No fixed abode,2025-02-02,Active,NaN,531.45,1,4,19
3,C-30003,Weybridge,2,60-74,Spanish,Owner-occupier,2024-10-06,Active,NaN,926.43,1,7,6
4,C-30004,Weybridge,1,30-44,English,No fixed abode,2023-03-19,Active,NaN,514.36,1,0,20



--- PAYMENTS HEAD ---


,payment_id,case_id,pay_month,amount,method,adjustment
0,P-123284,C-33951,2025-09,639.06,Transfer,N
1,P-118948,C-33218,2025-10,959.03,Transfer,N
2,P-103636,C-30614,2025-12,951.07,Transfer,N
3,P-111799,C-31999,2025-07,605.32,Transfer,N
4,P-105569,C-30943,2025-09,901.12,Transfer,N


## 6. Data Dictionary / Column Meanings

Based on the initial inspection and project context, here are the explanations of each column in the datasets:

### `cases.csv`
| Column Name | Data Type | Description |
| :--- | :--- | :--- |
| `case_id` | object / string | Unique identifier for each benefit case. |
| `district` | object / string | Geographic district of the client. |
| `household_size` | int64 | Number of people in the household. |
| `age_band` | object / string | Age category of the primary case holder. |
| `language_preference` | object / string | Preferred communication language (English, Spanish, Other). |
| `tenure` | object / string | Housing situation (e.g., Private tenancy, Social tenancy, No fixed abode). |
| `opened_date` | object / string | The date the benefit case was first opened (YYYY-MM-DD). |
| `status` | object / string | Current state of the case (Active, Closed, Suspended). |
| `closure_month` | object / string | Month in which the case was closed (YYYY-MM), empty (NaN) if not closed. |
| `monthly_award` | float64 | Standard monthly award amount allocated to the case. |
| `payment_adjustments` | int64 | Count of manual/historical benefit adjustments made to the case. |
| `contact_attempts` | int64 | Count of contact attempts made by the agency. |
| `months_since_review` | int64 | Number of months since the case was last reviewed. |

### `payments.csv`
| Column Name | Data Type | Description |
| :--- | :--- | :--- |
| `payment_id` | object / string | Unique identifier for each individual payment. |
| `case_id` | object / string | Identifier of the case receiving the payment (foreign key to `cases.csv`). |
| `pay_month` | object / string | Month when the payment was disbursed (YYYY-MM). |
| `amount` | float64 | The amount paid in this specific transaction. |
| `method` | object / string | Payment disbursement method (e.g., Transfer, Card). |
| `adjustment` | object / string | Indicator ('Y'/'N') of whether this transaction had an adjustment. |

## 7. Missing Values Check

We identify missing values (nulls) in both datasets.

In [6]:
print('--- MISSING VALUES IN CASES ---')
print(df_cases.isnull().sum())
print('\n--- MISSING VALUES IN PAYMENTS ---')
print(df_payments.isnull().sum())

--- MISSING VALUES IN CASES ---
case_id                   0
district                  0
household_size            0
age_band                  0
language_preference       0
tenure                    0
opened_date               0
status                    0
closure_month          3949
monthly_award             0
payment_adjustments       0
contact_attempts          0
months_since_review       0
dtype: int64

--- MISSING VALUES IN PAYMENTS ---
payment_id    0
case_id       0
pay_month     0
amount        0
method        0
adjustment    0
dtype: int64


## 8. Duplicate Rows and Duplicate IDs Check

We check for duplicate records in each file, as well as checking whether the unique ID keys (`case_id` in cases, `payment_id` in payments) contain duplicates.

In [7]:
print('--- DUPLICATE ROWS CHECK ---')
print('Duplicate rows in cases:', df_cases.duplicated().sum())
print('Duplicate rows in payments:', df_payments.duplicated().sum())

print('\n--- DUPLICATE KEY CHECKS ---')
print('Duplicate case_id in cases:', df_cases['case_id'].duplicated().sum())
print('Duplicate payment_id in payments:', df_payments['payment_id'].duplicated().sum())

--- DUPLICATE ROWS CHECK ---
Duplicate rows in cases: 0
Duplicate rows in payments: 0

--- DUPLICATE KEY CHECKS ---
Duplicate case_id in cases: 0
Duplicate payment_id in payments: 0


## 9. Unique Values in Categorical Columns

We list unique values and their value counts for each categorical variable in both files.

In [8]:
categorical_cols_cases = ['district', 'age_band', 'language_preference', 'tenure', 'status']
print('--- CASES CATEGORICAL UNIQUE VALUES ---')
for col in categorical_cols_cases:
    print(f'\nColumn: {col} ({df_cases[col].nunique()} unique values)')
    print(df_cases[col].value_counts())

categorical_cols_payments = ['method', 'adjustment']
print('\n--- PAYMENTS CATEGORICAL UNIQUE VALUES ---')
for col in categorical_cols_payments:
    print(f'\nColumn: {col} ({df_payments[col].nunique()} unique values)')
    print(df_payments[col].value_counts())

--- CASES CATEGORICAL UNIQUE VALUES ---

Column: district (4 unique values)
district
Calder Central    1433
Northgate         1121
Weybridge          983
Ash Hill           663
Name: count, dtype: int64

Column: age_band (5 unique values)
age_band
30-44    1112
45-59     992
60-74     874
18-29     701
75+       521
Name: count, dtype: int64

Column: language_preference (3 unique values)
language_preference
English    3277
Spanish     510
Other       413
Name: count, dtype: int64

Column: tenure (5 unique values)
tenure
Social tenancy          876
No fixed abode          855
Private tenancy         830
Resident with family    824
Owner-occupier          815
Name: count, dtype: int64

Column: status (3 unique values)
status
Active       3728
Closed        251
Suspended     221
Name: count, dtype: int64

--- PAYMENTS CATEGORICAL UNIQUE VALUES ---

Column: method (2 unique values)
method
Transfer    18520
Card         6236
Name: count, dtype: int64

Column: adjustment (2 unique values)
ad

## 10. Date Ranges Analysis

We inspect the time bounds of the datasets: `opened_date` and `closure_month` in cases, and `pay_month` in payments.

In [9]:
print('--- DATE RANGE CHECKS ---')
print(f"Cases opened_date range: {df_cases['opened_date'].min()} to {df_cases['opened_date'].max()}")
print(f"Cases closure_month range (non-nulls): {df_cases['closure_month'].dropna().min()} to {df_cases['closure_month'].dropna().max()}")
print(f"Payments pay_month range: {df_payments['pay_month'].min()} to {df_payments['pay_month'].max()}")

--- DATE RANGE CHECKS ---
Cases opened_date range: 2021-01-01 to 2025-08-28
Cases closure_month range (non-nulls): 2025-07 to 2025-12
Payments pay_month range: 2025-07 to 2025-12


## 11. Numeric Column Summary Statistics

We check descriptive summary statistics (mean, std, min, percentiles, max) for numeric fields in both datasets.

In [10]:
print('--- CASES NUMERIC COLUMNS SUMMARY ---')
display(df_cases.describe())
print('\n--- PAYMENTS NUMERIC COLUMNS SUMMARY ---')
display(df_payments.describe())

--- CASES NUMERIC COLUMNS SUMMARY ---


,household_size,monthly_award,payment_adjustments,contact_attempts,months_since_review
count,4200.000000,4200.000000,4200.000000,4200.000000,4200.000000
mean,2.469286,824.029710,0.969048,2.351190,11.504762
std,1.365903,279.527184,0.985967,1.687619,6.830962
min,1.000000,372.150000,0.000000,0.000000,0.000000
25%,1.000000,612.580000,0.000000,1.000000,6.000000
50%,2.000000,765.420000,1.000000,2.000000,11.000000
75%,3.000000,1000.122500,2.000000,3.000000,17.000000
max,6.000000,1850.800000,6.000000,9.000000,23.000000



--- PAYMENTS NUMERIC COLUMNS SUMMARY ---


,amount
count,24756.000000
mean,812.886067
std,288.040953
min,253.750000
25%,597.687500
50%,757.475000
75%,988.347500
max,2729.070000


## 12. Verify Relationship between Cases and Payments

We check how the datasets align on the primary/foreign key `case_id`.

In [11]:
cases_ids = set(df_cases['case_id'])
payments_ids = set(df_payments['case_id'])
print(f'Unique cases in cases.csv: {len(cases_ids)}')
print(f'Unique cases in payments.csv: {len(payments_ids)}')
print(f'Intersection of cases: {len(cases_ids.intersection(payments_ids))}')

Unique cases in cases.csv: 4200
Unique cases in payments.csv: 4200
Intersection of cases: 4200


## 13. Payment Count Distribution per Case

We count the number of payments associated with each case and examine the distribution.

In [12]:
payments_per_case = df_payments.groupby('case_id').size()
print('Summary Statistics of Payments per Case:')
print(payments_per_case.describe())

# Value counts of number of payments
print('\nFrequency of payment counts:')
print(payments_per_case.value_counts().sort_index())

Summary Statistics of Payments per Case:
count    4200.000000
mean        5.894286
std         0.671444
min         1.000000
25%         6.000000
50%         6.000000
75%         6.000000
max        10.000000
dtype: float64

Frequency of payment counts:
1       32
2       35
3       34
4       35
5       33
6     3986
7       32
8       11
9        1
10       1
Name: count, dtype: int64


## 14. Orphaned Cases Check (Cases with No Payments)

We verify if there are any cases that do not have a single corresponding transaction in `payments.csv`.

In [13]:
orphaned_cases = cases_ids - payments_ids
print(f'Number of cases with no payment record: {len(orphaned_cases)}')
if len(orphaned_cases) > 0:
    print('Sample orphaned cases:', list(orphaned_cases)[:5])

Number of cases with no payment record: 0


## 15. Orphaned Payments Check (Payments for Missing Cases)

We verify if there are any payments in `payments.csv` referring to a `case_id` that is absent from `cases.csv`.

In [14]:
orphaned_payments = payments_ids - cases_ids
print(f'Number of payments referencing a missing case: {len(orphaned_payments)}')
if len(orphaned_payments) > 0:
    print('Sample orphaned payments:', list(orphaned_payments)[:5])

Number of payments referencing a missing case: 0


## 16. Relevant Fields for Fairness Analysis

Fairness analysis requires evaluating model performance or outcome parity across different demographic or socioeconomic subgroups.
Based on `cases.csv`, the following columns are key demographic/population indicators that will be relevant for fairness analysis later:

1. **`district`**: Geographic region of the client. Important to test for locational equity.
2. **`age_band`**: Age range of the client (18-29, 30-44, 45-59, 60-74, 75+). Important to check for age discrimination or bias.
3. **`language_preference`**: Communication preference (English, Spanish, Other). Acts as a potential proxy for nationality, race, or language proficiency.
4. **`tenure`**: Housing situation (Private tenancy, Social tenancy, Owner-occupier, Resident with family, No fixed abode). Serves as a socioeconomic proxy variable.

## 17. Absence of Target Label

> [!IMPORTANT]
> **Note on Target Variable:** There is **no explicit target label** (e.g. `is_improper`, `overpaid`, `fraudulent`) in either `cases.csv` or `payments.csv` identifying improper payments.
>
> This means our task will require deriving indicators or proxy metrics for potential overpayment, or building an unsupervised/semi-supervised prioritization model rather than a standard supervised classification model.

## Initial Observations

The following observations are directly supported by the data profiling and checking:

1. **Case Status Distribution**: The dataset contains 4,200 benefit cases, where **3,728 (88.76%)** are `Active`, **251 (5.98%)** are `Closed`, and **221 (5.26%)** are `Suspended`.
2. **Payment Temporal Scope**: Payments are tracked for a 6-month window from **2025-07 to 2025-12** (inclusive).
3. **Perfect Case-Payment Alignment**: There is a 1-to-1 linkage on `case_id` between `cases.csv` and `payments.csv`. No cases lack payments, and no payments refer to non-existent cases.
4. **Payments for Closed Cases (Potential Overpayment Signal)**:
   * Cases marked as `Closed` have an average of ~4 payments. However, examining payment dates against `closure_month` reveals that some cases continued to receive payments *after* their closure month.
   * For example, out of cases closed in `2025-08`, some cases continued to receive payments in `2025-09` through `2025-12`. This represents a strong, objective data signal for potential improper overpayments.
5. **Duplicate Case-Month Transactions**: There are 62 instances of a case receiving multiple payment records in a single month (often using different payment methods such as one via Transfer and one via Card for the same amount). This could represent system duplication errors or split disbursements.
6. **Discrepancy Between Awards and Payments**: The payment amount (`amount` in payments) and the declared monthly benefit amount (`monthly_award` in cases) differ in the vast majority of transactions. Only a small fraction (343 out of 24,756) match within $0.01. Further analysis is required to understand if these differences are due to adjustments or other variables.
7. **Fairness Demographics**: District, Age Band, Language Preference, and Housing Tenure have complete data (0% missing) for all 4,200 cases, which will allow for robust demographic parity checks during fairness evaluation.